### PR TITLE
fix: [#60] Fetch messages from homepage via https

### LINKED ISSUE #60: Validate the MessageParser works correctly
In the logs, I see the following Fatal message that seems to come from parsing central messages:

<pre>
INFO [tr.messages]: Last message check date: 1/1/70, 1:00 AM
INFO [null]: Last record repeated again.
[Fatal Error] :1:1: Premature end of file.
INFO [tr.messages]: Finished
</pre>

Ensure the code is robust and the message is syntactically correct and write an integration test to document this and decrease the risk of regression

### PR DESCRIPTION
**Pull request by ursjoss**
Fixes #60.

- Changes the URL for the message file from http to https, fixing the retrieval in the `MessageChecker` (more to come in #59).
- Exercise the `MessageChecker` and the `MessageParser` via some new unit tests
- Add some `@NonNull` annotations to both classes.

### CODE REVIEW

File: modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import java.util.List;
+import static org.junit.Assert.*;
+import org.junit.Test;
+
+public class MessageParserTest {
+
+    private MessageParser parser = new MessageParser();
+    private LocalDate localDate = LocalDate.of(2014, Month.MARCH, 26);
+
+    final String xml1 = "<msgs>"
```
Comment (RadekCap) [inline modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java:19]: I would prefer sampleMessage1 or sampleXmlMessage1. xml1 seems like a little bit cryptic.

File: modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            + "<msg date='2014/03/30' type='M'>memberMessage</msg>"
+            + "</msgs>";
+
+    @Test
+    public void parsingValidButEmptyXml_returnsNoMessages() throws Exception {
+        assertTrue(parse("<msgs/>", localDate).isEmpty());
+    }
+
+    @Test
+    public void parsingXml1_withDateBeforeBoth_returnsBothMessages() throws Exception {
```
Comment (RadekCap) [inline modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java:35]: In dependency of provious change this should be
parsingSampleMessage1_witchDateBeforeBoth_returnsBoothMessages 
etc.

File: modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+import java.time.LocalDate;
+import java.time.Month;
+import java.util.List;
+import static org.junit.Assert.*;
+import org.junit.Test;
+
+public class MessageParserTest {
+
+    private MessageParser parser = new MessageParser();
+    private LocalDate localDate = LocalDate.of(2014, Month.MARCH, 26);
```
Comment (RadekCap) [inline modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java:17]: Why this date? What about introduce some variable with proper name explaining why this date?

File: modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    @Test
+    public void parsingValidButEmptyXml_returnsNoMessages() throws Exception {
+        assertTrue(parse("<msgs/>", localDate).isEmpty());
+    }
+
+    @Test
+    public void parsingXml1_withDateBeforeBoth_returnsBothMessages() throws Exception {
+        List<Message> messages = parse(xml1, localDate);
+        assertTrue(messages.size() == 2);
+        assertMessage(messages.get(0), LocalDate.of(2014, Month.MARCH, 27), "a", "message1");
```
Comment (RadekCap) [inline modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java:38]: This is everywhere. And, the second date as well. Extract them to the variable?

File: modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+    @Test
+    public void parseFile() throws Exception {
+        Path path = Paths.get(getClass().getResource("messages.xml").getPath());
+        String xml = new String(Files.readAllBytes(path));
+        List<Message> messages = parse(xml, localDate);
+        assertTrue(messages.size() == 1);
+        assertMessage(messages.get(0), LocalDate.of(2014, Month.MARCH, 27), "a", null);
+        String text = messages.get(0).text;
+        assertEquals(431, text.length());
```
Comment (RadekCap) [inline modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java:73]: 431? What does it mean? Why 431? Probably the length of the original message?
I would prefer variable like
const int LENGHT_OF_SAMPLE_MESSAGE  = 431;

File: modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/messages.xml [hunk 6]
```diff
@@ -0,0 +1,122 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (RadekCap) [inline modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/messages.xml:1]: It's good to know where is the changelog 👍

**Review by RadekCap — CHANGES_REQUESTED** [review]
Nice changes.

You can leave it as is and merge it, it's good to go.

My personal preferences are implement suggested changes related to naming. I leave it up to you. I leave it up to you.

Thank you!

P.S.: I thought I'll be doing some development here, but you started in a pace where I barely do PR reviews 👍

**Comment by ursjoss** [comment]
@RadekCap Thanks for the thorogh review. I hope I addressed all your issues. Please recheck.

File: modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java (hunk 2 above)
Comment (ursjoss) [reply modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java:47]: Had to get off the train, before i could reread all. Will follow up, thanks

**Comment by ursjoss** [comment]
@RadekCap I set the `MessageCheckerTest` to be ignored by default. It's rather for ad-hoc-tests and will fail after a successful first run.

File: modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+public class MessageParserTest {
+
+    // The message text stored in the sample messages.xml (pulled from the
+    // https://trgtd.com.au/messages.xml on 2023-01-14) is too long and
+    // cumbersome to validate verbatim.Simply verifying the length instead.
+    private static final int LENGTH_OF_SAMPLE_MESSAGE_TEXT = 431;
+
+    private final MessageParser parser = new MessageParser();
+
+    private final LocalDate dateMessage1 = LocalDate.of(2014, Month.MARCH, 27);
```
Comment (RadekCap) [inline modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java:24]: dateMessage1, dateMessage2 - it's hard to read what the context of this dates.
This should be probably firstSampleDate and secondSampleDate.

**Review by RadekCap — CHANGES_REQUESTED** [review]
Just one small change for you, and re-run of check and check of passing newly added tests for me.

File: modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java (hunk 7 above)
Comment (ursjoss) [reply modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java:24]: @RadekCap Not sure if I agree with your wording. Let me quickly explain my understanding:

Conceptually, we have an xml document with a set of messages. Each message has a date, a type, a text and an href.

Currently, in the MessageParserTest, we have two samples of such an xml document:

- sampleMessagesA: We use it to specify that depending on the date we use for parsing, we get both, none (not actually tested, but I might add that case too) or one of those messages.
- sampleMessagesB: We use it to differentiate between member users and non-member users only retrieving the messages intended for each group.

For simplicity, the first of the two messages of both samples are assigned one date (dateMessage1), the second of message of each sample has another date (dateMessage2). We could call it dateOfMessage2 if you prefer, but imho firstSampleDate has the wrong semantic.

Do you disagree?

Comment (RadekCap) [reply modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java]: np, that's ok as is.

Comment (ursjoss) [reply modules/au.com.trgtd.tr.task.messages/test/unit/src/au/com/trgtd/tr/task/messages/MessageParserTest.java]: Excellent, I'll merge then. Thanks for the thorough review. Very good improvements coming from it.

**Review by RadekCap — APPROVED** [review]
MessageParserTests are passing on CI, good to go.

Nice work, thank you :)

**Pull request merged**